File: sih26168_v2x/python/sih26168_v2x/frames.py
from __future__ import annotations

import math

from .types import EARTH_RADIUS_M, EnuOrigin, EnuVector, GeoPosition


def finite_number(x: float) -> bool:
    return math.isfinite(x)


def valid_latitude(lat_deg: float) -> bool:
    return finite_number(lat_deg) and -90.0 <= lat_deg <= 90.0


def valid_longitude(lon_deg: float) -> bool:
    return finite_number(lon_deg) and -180.0 <= lon_deg <= 180.0


def wrap_pi(rad: float) -> float:
    if not math.isfinite(rad):
        return rad
    return math.atan2(math.sin(rad), math.cos(rad))


def geo_to_local_enu(geo: GeoPosition, origin: EnuOrigin) -> EnuVector:
    if (
        not origin.valid
        or not valid_latitude(geo.latitude_deg)
        or not valid_longitude(geo.longitude_deg)
        or not valid_latitude(origin.latitude_deg)
        or not valid_longitude(origin.longitude_deg)
        or not finite_number(geo.altitude_m)
        or not finite_number(origin.altitude_m)
    ):
        return EnuVector()
    lat0 = math.radians(origin.latitude_deg)
    north = math.radians(geo.latitude_deg - origin.latitude_deg) * EARTH_RADIUS_M
    east = math.radians(geo.longitude_deg - origin.longitude_deg) * EARTH_RADIUS_M * math.cos(lat0)
    return EnuVector(east_m=east, north_m=north, up_m=geo.altitude_m - origin.altitude_m)


def local_enu_to_geo(enu: EnuVector, origin: EnuOrigin) -> GeoPosition:
    geo = GeoPosition(origin.latitude_deg, origin.longitude_deg, origin.altitude_m)
    if not origin.valid:
        return geo
    lat0 = math.radians(origin.latitude_deg)
    c = max(abs(math.cos(lat0)), 1e-8)
    geo.latitude_deg = origin.latitude_deg + math.degrees(enu.north_m / EARTH_RADIUS_M)
    geo.longitude_deg = origin.longitude_deg + math.degrees(enu.east_m / (EARTH_RADIUS_M * c))
    geo.altitude_m = origin.altitude_m + enu.up_m
    return geo


def enu_to_member3_north_east(enu: EnuVector) -> tuple[float, float]:
    return enu.north_m, enu.east_m


def member3_north_east_to_enu(north_m: float, east_m: float, up_m: float = 0.0) -> EnuVector:
    return EnuVector(east_m=east_m, north_m=north_m, up_m=up_m)


def vehicle_frame_to_navigation_enu(v_x: float, v_y: float, yaw_rad: float) -> tuple[float, float]:
    """Member 3 predict() kinematics: yaw 0 = North, increasing toward East."""
    c, s = math.cos(yaw_rad), math.sin(yaw_rad)
    v_north = v_x * c - v_y * s
    v_east = v_x * s + v_y * c
    return v_east, v_north


def heading_speed_to_enu(heading_rad: float, speed_mps: float) -> tuple[float, float]:
    v_north = speed_mps * math.cos(heading_rad)
    v_east = speed_mps * math.sin(heading_rad)
    return v_east, v_north

File: sih26168_v2x/src/frames.cpp
#include "sih26168/v2x/frames.hpp"

#include <cmath>
#include <limits>
#include <numbers>

namespace sih26168::v2x {
namespace {

constexpr double kPi = std::numbers::pi;

}  // namespace

bool finiteNumber(double x) {
    return std::isfinite(x);
}

bool validLatitude(double lat_deg) {
    return finiteNumber(lat_deg) && lat_deg >= -90.0 && lat_deg <= 90.0;
}

bool validLongitude(double lon_deg) {
    return finiteNumber(lon_deg) && lon_deg >= -180.0 && lon_deg <= 180.0;
}

double wrapPi(double rad) {
    if (!std::isfinite(rad)) {
        return rad;
    }
    return std::atan2(std::sin(rad), std::cos(rad));
}

EnuVector geoToLocalENU(const GeoPosition& geo, const EnuOrigin& origin) {
    EnuVector enu{};
    if (!origin.valid || !validLatitude(geo.latitude_deg) || !validLongitude(geo.longitude_deg) ||
        !validLatitude(origin.latitude_deg) || !validLongitude(origin.longitude_deg) ||
        !finiteNumber(geo.altitude_m) || !finiteNumber(origin.altitude_m)) {
        return enu;
    }
    const double lat0 = origin.latitude_deg * kPi / 180.0;
    enu.north_m = (geo.latitude_deg - origin.latitude_deg) * kPi / 180.0 * kEarthRadiusM;
    enu.east_m = (geo.longitude_deg - origin.longitude_deg) * kPi / 180.0 * kEarthRadiusM *
                 std::cos(lat0);
    enu.up_m = geo.altitude_m - origin.altitude_m;
    return enu;
}

GeoPosition localENUToGeo(const EnuVector& enu, const EnuOrigin& origin) {
    GeoPosition geo{};
    geo.latitude_deg = origin.latitude_deg;
    geo.longitude_deg = origin.longitude_deg;
    geo.altitude_m = origin.altitude_m;
    if (!origin.valid || !finiteNumber(enu.east_m) || !finiteNumber(enu.north_m) ||
        !finiteNumber(enu.up_m)) {
        return geo;
    }
    const double lat0 = origin.latitude_deg * kPi / 180.0;
    geo.latitude_deg = origin.latitude_deg + enu.north_m / kEarthRadiusM * 180.0 / kPi;
    const double c = std::max(std::abs(std::cos(lat0)), 1e-8);
    geo.longitude_deg = origin.longitude_deg + enu.east_m / (kEarthRadiusM * c) * 180.0 / kPi;
    geo.altitude_m = origin.altitude_m + enu.up_m;
    return geo;
}

void enuToMember3NorthEast(const EnuVector& enu, double& north_m, double& east_m) {
    north_m = enu.north_m;
    east_m = enu.east_m;
}

EnuVector member3NorthEastToEnu(double north_m, double east_m, double up_m) {
    return EnuVector{east_m, north_m, up_m};
}

void vehicleFrameToNavigationEnu(double v_x, double v_y, double yaw_rad, double& v_east_mps,
                                 double& v_north_mps) {
    const double c = std::cos(yaw_rad);
    const double s = std::sin(yaw_rad);
    v_north_mps = v_x * c - v_y * s;
    v_east_mps = v_x * s + v_y * c;
}

void navigationEnuToVehicleFrame(double v_east_mps, double v_north_mps, double yaw_rad, double& v_x,
                                 double& v_y) {
    const double c = std::cos(yaw_rad);
    const double s = std::sin(yaw_rad);
    v_x = v_north_mps * c + v_east_mps * s;
    v_y = -v_north_mps * s + v_east_mps * c;
}

void headingSpeedToEnu(double heading_rad, double speed_mps, double& v_east, double& v_north) {
    v_north = speed_mps * std::cos(heading_rad);
    v_east = speed_mps * std::sin(heading_rad);
}

}  // namespace sih26168::v2x
